### PR TITLE
AIP-72: Adding PUT Variable Endpoint for execution API

### LINKED ISSUE #44446: Add "Post variable" endpoint for execution API
### Body

Similar to https://github.com/apache/airflow/pull/43832 but for setting a variable

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

### PR DESCRIPTION
**Pull request by amoghrajesh**
closes: https://github.com/apache/airflow/issues/44446
Adding an API to the execution API for "setting" or "creating" a variable, later to be used by task SDK

Examples from legacy API and this one:

1. Getting one variable using Legacy API:
image

2. Getting one variable using execution API:
image

3. Setting a variable using legacy API:
image

4. Setting a variable using execution API:
image

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/api_fastapi/execution_api/routes/variables.py [hunk 1]
```diff
@@ -67,6 +67,20 @@ def get_variable(variable_key: str, token: deps.TokenDep) -> VariableResponse:
...
+@router.post(
+    "",
+    responses={
+        status.HTTP_401_UNAUTHORIZED: {"description": "Unauthorized"},
+        status.HTTP_403_FORBIDDEN: {"description": "Task does not have access to the variable"},
+    },
+)
+def post_variable(body: VariablePostBody, token: deps.TokenDep) -> VariableResponse:
+    """Set an Airflow Variable."""
+    # TODO: add the token validation
```
Comment (amoghrajesh) [inline airflow/api_fastapi/execution_api/routes/variables.py]: I am yet to handle this part, wasn't too sure if I should follow the current template or ignore for now

Comment (ashb) [reply airflow/api_fastapi/execution_api/routes/variables.py]: Call `has_variable_access`, and add a new parameter to that function to make it `has_variable_access(variable_key, token, write=True)` etc.

Though we should probably make better use of decorators/dependencies so we don't need to handle this _inside_ each view function, but for now just follow the pattern.

File: tests/api_fastapi/execution_api/routes/test_variables.py [hunk 2]
```diff
@@ -23,10 +23,21 @@
...
+class TestVariableEndpoint:
+    @pytest.fixture(autouse=True)
+    def setup_method(self) -> None:
+        clear_db_variables()
+
+    def teardown_method(self) -> None:
+        clear_db_variables()
+
+
+class TestGetVariable(TestVariableEndpoint):
```
Comment (ashb) [inline tests/api_fastapi/execution_api/routes/test_variables.py]: I don't think we need a base class for this -- these can be done like this:

```suggestion
@pytest.fixture(autouse=True)
def setup_method(self) -> None:
    clear_db_variables()
    yield
    clear_db_variables()


class TestGetVariable::
```

File: tests/api_fastapi/execution_api/routes/test_variables.py [hunk 3]
```diff
@@ -75,3 +86,60 @@ def test_variable_get_access_denied(self, client):
...
+
+class TestPostVariable(TestVariableEndpoint):
+    def test_should_create_variable(self, client, session):
+        payload = {"key": "var_create", "value": "{}"}
+        response = client.post(
+            "/execution/variables",
+            json=payload,
+        )
+        assert response.status_code == 200
+        response = client.get(f"/execution/variables/{payload['key']}")
```
Comment (ashb) [inline tests/api_fastapi/execution_api/routes/test_variables.py]: This is implicitly testing the `get` resposne too. Should we perhaps instead go directly to the DB and test what is in there?

File: tests/api_fastapi/execution_api/routes/test_variables.py [hunk 4]
```diff
@@ -75,3 +86,60 @@ def test_variable_get_access_denied(self, client):
...
+            "key": "var_create",
+            "value": "{}",
+        }
+
+    @pytest.mark.parametrize(
+        "payload, status_code",
+        [
+            pytest.param(
+                {"key": "var_create", "value": "{}", "description": "description"}, 200, id="valid-payload"
+            ),
```
Comment (ashb) [inline tests/api_fastapi/execution_api/routes/test_variables.py]: I'd say remove this -- we're already testing it via `test_should_create_variable` and having a success case in here complicates the test.

File: tests/api_fastapi/execution_api/routes/test_variables.py [hunk 5]
```diff
@@ -75,3 +86,60 @@ def test_variable_get_access_denied(self, client):
...
+            pytest.param({"value": "{}", "description": "description"}, 422, id="missing-key"),
+            pytest.param({"key": "var_create", "description": "description"}, 422, id="missing-value"),
+            pytest.param({"value": "{}", "key": "var_create"}, 200, id="missing-description"),
```
Comment (ashb) [inline tests/api_fastapi/execution_api/routes/test_variables.py]: Ditto here. If you want to test multiple success cases move make the "should create" test parameterized.

File: airflow/api_fastapi/execution_api/routes/variables.py [hunk 6]
```diff
@@ -67,6 +67,20 @@ def get_variable(variable_key: str, token: deps.TokenDep) -> VariableResponse:
...
 
 
+@router.post(
+    "",
+    responses={
+        status.HTTP_401_UNAUTHORIZED: {"description": "Unauthorized"},
+        status.HTTP_403_FORBIDDEN: {"description": "Task does not have access to the variable"},
+    },
+)
+def post_variable(body: VariablePostBody, token: deps.TokenDep) -> VariableResponse:
```
Comment (ashb) [inline airflow/api_fastapi/execution_api/routes/variables.py]: Design question: Should this be `POST /execution/variables` with key in the body, or should it perhaps be `PUT /execution/variables/{variable_key}`?

Possible advantage of the second approach: it shows in the request logs when a specific variable is updated which might be nice.

Ultimately it doesn't really matter, and there is no "correct" way, it's just up to us.

Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/variables.py]: Right.

Two reasons i chose POST `/execution/variables` over the other one:
1. `variable_key` and `key` again in the payload didn't feel too right as it was avoidable duplication.
2. We have to handle cases where `variable_key` in the parameters and `key` in the payload are potentially different. Could be quite chaotic!

Also, legacy implements it this way!

Comment (ashb) [reply airflow/api_fastapi/execution_api/routes/variables.py]: On 1. Yeah, if it's in the URL we wouldn't have it in the body too, just value and desc.

Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/variables.py]: I see, it doesn't really matter to me which way we go then. Just a matter of agreement with everyone

File: airflow/api_fastapi/execution_api/routes/variables.py (hunk 1 above)
Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/variables.py]: Sure, did this: 
```
    if not has_variable_access(body.key, token, write_access=True):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail={
                "reason": "access_denied",
                "message": f"Task does not have access to write variable {body.key}",
            },
        )

```

File: tests/api_fastapi/execution_api/routes/test_variables.py (hunk 2 above)
Comment (amoghrajesh) [reply tests/api_fastapi/execution_api/routes/test_variables.py]: Handled, it doesn't have a `self`

File: tests/api_fastapi/execution_api/routes/test_variables.py (hunk 3 above)
Comment (amoghrajesh) [reply tests/api_fastapi/execution_api/routes/test_variables.py]: Yeah that makes sense. This is testing another API

Comment (amoghrajesh) [reply tests/api_fastapi/execution_api/routes/test_variables.py]: Handled the change

File: tests/api_fastapi/execution_api/routes/test_variables.py (hunk 4 above)
Comment (amoghrajesh) [reply tests/api_fastapi/execution_api/routes/test_variables.py]: Yeah, makes sense. removing

File: tests/api_fastapi/execution_api/routes/test_variables.py (hunk 5 above)
Comment (amoghrajesh) [reply tests/api_fastapi/execution_api/routes/test_variables.py]: Added test cases to `test_should_create_variable` instead. Good idea

File: airflow/api_fastapi/execution_api/routes/variables.py [hunk 7]
```diff
@@ -67,7 +67,29 @@ def get_variable(variable_key: str, token: deps.TokenDep) -> VariableResponse:
...
+        raise HTTPException(
+            status_code=status.HTTP_403_FORBIDDEN,
+            detail={
+                "reason": "access_denied",
+                "message": f"Task does not have access to write variable {body.key}",
+            },
+        )
+
+    Variable.set(**body.model_dump())
+    return VariableResponse(key=body.key, value=body.value)
```
Comment (kaxil) [inline airflow/api_fastapi/execution_api/routes/variables.py]: We don't need to send the key and value back as a response.

Comment (kaxil) [reply airflow/api_fastapi/execution_api/routes/variables.py]: Just `status.HTTP_201_CREATED`

File: task_sdk/tests/execution_time/test_supervisor.py [hunk 8]
```diff
@@ -320,30 +314,30 @@ def watched_subprocess(self, mocker):
...
+            #     id="get_variable",
+            # ),
+            # pytest.param(
+            #     GetXCom(dag_id="test_dag", run_id="test_run", task_id="test_task", key="test_key"),
+            #     b'{"key":"test_key","value":"test_value"}\n',
+            #     "xcoms.get",
+            #     ("test_dag", "test_run", "test_task", "test_key", -1),
+            #     XComResult(key="test_key", value="test_value"),
+            #     id="get_xcom",
+            # ),
```
Comment (kaxil) [inline task_sdk/tests/execution_time/test_supervisor.py]: Unrelated?

Comment (amoghrajesh) [reply task_sdk/tests/execution_time/test_supervisor.py]: Ah yes, was working on a different branch and this went in :/

File: airflow/api_fastapi/execution_api/routes/variables.py (hunk 7 above)
Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/variables.py]: Yea should be good enough, but I wasn't too sure why the legacy and the new core api returns the variable.
But in our case, should be ok

Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/variables.py]: Made the changes

File: airflow/api_fastapi/execution_api/routes/variables.py (hunk 6 above)
Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/variables.py]: Based on offline discussion we decided to go with `PUT /execution/variables/{variable_key}`

File: tests/api_fastapi/execution_api/routes/test_variables.py [hunk 9]
```diff
@@ -75,3 +84,81 @@ def test_variable_get_access_denied(self, client):
...
+        assert response.status_code == status_code
+        if response.status_code == 422:
+            assert response.json()["detail"][0]["type"] == "missing"
+            assert response.json()["detail"][0]["msg"] == "Field required"
+
+    def test_overwriting_existing_variable(self, client, session):
+        Variable.set(key="var_create", value="value", session=session)
+        session.commit()
+
+        key = "var_create"
```
Comment (ashb) [inline tests/api_fastapi/execution_api/routes/test_variables.py]: Nit:
```suggestion
        key = "var_create"
        Variable.set(key=key, value="value", session=session)
        session.commit()

```

File: tests/api_fastapi/execution_api/routes/test_variables.py [hunk 10]
```diff
@@ -75,3 +84,81 @@ def test_variable_get_access_denied(self, client):
...
+        payload = {"value": "new_value"}
+        response = client.put(
+            f"/execution/variables/{key}",
+            json=payload,
+        )
+        assert response.status_code == 201
+        # variable should have been updated to the new value
+        var_from_db = session.query(Variable).where(Variable.key == "var_create").first()
```
Comment (ashb) [inline tests/api_fastapi/execution_api/routes/test_variables.py]:
```suggestion
        var_from_db = session.query(Variable).where(Variable.key == key).first()
```

**Review by ashb — APPROVED** [review]
Two small nits, LGTM

File: tests/api_fastapi/execution_api/routes/test_variables.py (hunk 10 above)
Comment (amoghrajesh) [reply tests/api_fastapi/execution_api/routes/test_variables.py]: Ah thanks

File: tests/api_fastapi/execution_api/routes/test_variables.py (hunk 9 above)
Comment (amoghrajesh) [reply tests/api_fastapi/execution_api/routes/test_variables.py]: Done

File: airflow/api_fastapi/execution_api/routes/variables.py [hunk 11]
```diff
@@ -67,7 +67,28 @@ def get_variable(variable_key: str, token: deps.TokenDep) -> VariableResponse:
...
 
-def has_variable_access(variable_key: str, token: TIToken) -> bool:
+@router.put(
+    "/{variable_key}",
+    responses={
+        status.HTTP_401_UNAUTHORIZED: {"description": "Unauthorized"},
+        status.HTTP_403_FORBIDDEN: {"description": "Task does not have access to the variable"},
+    },
+)
+def post_variable(variable_key: str, body: VariablePostBody, token: deps.TokenDep):
```
Comment (kaxil) [inline airflow/api_fastapi/execution_api/routes/variables.py]:
```suggestion
def put_variable(variable_key: str, body: VariablePostBody, token: deps.TokenDep):
```

or just `variable_set` ?

File: airflow/api_fastapi/execution_api/routes/variables.py [hunk 12]
```diff
@@ -67,7 +67,28 @@ def get_variable(variable_key: str, token: deps.TokenDep) -> VariableResponse:
     return VariableResponse(key=variable_key, value=variable_value)
 
 
-def has_variable_access(variable_key: str, token: TIToken) -> bool:
+@router.put(
+    "/{variable_key}",
+    responses={
```
Comment (kaxil) [inline airflow/api_fastapi/execution_api/routes/variables.py:73]:
```suggestion
    status_code=status.HTTP_201_CREATED
    responses={
```

File: airflow/api_fastapi/execution_api/routes/variables.py [hunk 13]
```diff
@@ -67,7 +67,28 @@ def get_variable(variable_key: str, token: deps.TokenDep) -> VariableResponse:
...
+    if not has_variable_access(variable_key, token, write_access=True):
+        raise HTTPException(
+            status_code=status.HTTP_403_FORBIDDEN,
+            detail={
+                "reason": "access_denied",
+                "message": f"Task does not have access to write variable {variable_key}",
+            },
+        )
+    Variable.set(key=variable_key, value=body.value, description=body.description)
+    return Response(status_code=status.HTTP_201_CREATED)
```
Comment (kaxil) [inline airflow/api_fastapi/execution_api/routes/variables.py]:
```suggestion
    return {"message": "Variable successfully set"}
```

**Review by kaxil — APPROVED** [review]
lgtm, minor comments

File: airflow/api_fastapi/execution_api/routes/variables.py (hunk 11 above)
Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/variables.py]: `put_variable` sounds better imo. Adding for now.

File: airflow/api_fastapi/execution_api/routes/variables.py (hunk 12 above)
Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/variables.py:73]: Nice, handled!

File: airflow/api_fastapi/execution_api/routes/variables.py (hunk 13 above)
Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/variables.py]: Cool, handled!
